File: euler_060.py
from time import time
from math import sqrt
from primesieve import Iterator
from primePy import primes



def is_prime(n):
    it = Iterator()
    p = it.next_prime()

    lim = int(sqrt(n)) + 1
    while p < lim:
        if n % p == 0:
            return False
        p = it.next_prime()
    return True


def is_prime_pair(p, q):
    return is_prime(int(str(p) + str(q))) and is_prime(int(str(q) + str(p)))


start = time()

p_list = list(primes.upto(10000))
l = len(p_list)

pairs = {}
for p in p_list:
    pairs[p] = set()

for p in p_list:
    for q in p_list:
        if is_prime_pair(p, q):
            pairs[p].add(q)
            pairs[q].add(p)

candidates = [p for p in pairs if len(pairs[p]) > 4]

threes = set()

for a in candidates:
    for b in pairs[a]:
        for c in candidates:
            if c in pairs[a].intersection(pairs[b]):
                threes.add((a,b,c))

fours = set()

for t in threes:
    for c in candidates:
        if c in t:
            continue
        else:
            if c in pairs[t[0]].intersection(pairs[t[1]]).intersection(pairs[t[2]]):
                fours.add((t[0], t[1], t[2], c))

fives = set()

for q in fours:
    for c in candidates:
        if c in q:
            continue
        else:
            if c in pairs[q[0]].intersection(pairs[q[1]]).intersection(pairs[q[2]]).intersection(pairs[q[3]]):
                fives.add((q[0], q[1], q[2], q[3], c))

lowest_sum = 50000
lowest = (0,0,0,0,0)

for f in fives:
    if sum(f) < lowest_sum:
        lowest_sum = sum(f)
        lowest = f

print("Optimal set of  five: " + str(lowest) + ", sum: " + str(lowest_sum))



stop = time()
print("Time: " + str(stop - start))

File: euler_060.js
const start = Date.now();

// Primes
const upperBound = 10000;
let sieve = Array(upperBound);
sieve.fill(true); sieve[0] = false; sieve[1] = false;
let primes = [];

for (let i = 0; i < upperBound; i++) {
  if (sieve[i]) {
    primes.push(i);
    let m = 2;
    while (m * i < upperBound) {
      sieve[m * i] = false;
      m++;
    }
  }
}

const boundedPrimes = (bound) => {
  return primes.filter(x => x < bound);
}

const isPrime = (c) => {
  if (c < upperBound) {
    return sieve[c];
  }
  for (const bp of boundedPrimes(Math.ceil(Math.sqrt(c)))) {
    if (c % bp === 0) {
      return false;
    }
  }
  return true;
}

console.log("Primes generated. Number of primes is ", primes.length);
// End of prime generation

// Concatenation of primes and prime pairs

const concatenate = (p1, p2) => {
  return Number.parseInt(`${p1}${p2}`);
}

const isPrimePair = (p1, p2) => {
  return isPrime(concatenate(p1, p2)) && isPrime(concatenate(p2, p1));
}

const largerPrimePairs = (p) => {
  let res = [];
  for (const lp of primes.filter(x => x > p)) {
    if (isPrimePair(p, lp)) {
      res.push(lp);
    }
  }
  return res;
}

let primePairMap = new Map();

for (const p of primes) {
  const lpp = largerPrimePairs(p);
  if (lpp.length > 0) {
    primePairMap.set(p, lpp);
  }
}

const candidates = primes.filter(x => primePairMap.has(x) && primePairMap.get(x).length > 4);

let pairs = [];
for (const c of candidates) {
  for (const lp of primePairMap.get(c)) {
    if (primePairMap.has(lp)) {
      pairs.push([c, lp]);
    }
  }
}

let triples = [];
for (const pair of pairs) {
  for (const lp of primePairMap.get(pair[1])) {
    if (primePairMap.has(lp) && isPrimePair(pair[0], lp)) {
      triples.push([pair[0], pair[1], lp]);
    }
  }
}

let quads = [];
for (const triple of triples) {
  for (const lp of primePairMap.get(triple[2])) {
    if (primePairMap.has(lp) && triple.slice(0,2).every(x => isPrimePair(x, lp))) {
      quads.push([triple[0], triple[1], triple[2], lp]);
    }
  }
}

let quints = [];
for (const quad of quads) {
  for (const lp of primePairMap.get(quad[3])) {
    if (primePairMap.has(lp) && quad.slice(0,3).every(x => isPrimePair(x, lp))) {
      quints.push([quad[0], quad[1], quad[2], quad[3], lp]);
    }
  }
}


const sum = quints[0].reduce((p,c) => p + c,0);

const end = Date.now();

console.log("Execute time: ", end - start, " Lowest sum of five primes is: ", sum);